File: 0673-number-of-longest-increasing-subsequence/0673-number-of-longest-increasing-subsequence.java
// class Solution {
//     public int findNumberOfLIS(int[] nums) {
//         int n = nums.length;
//         int[] LIS = new int[n];

//         LIS[0] = 1;
//         int maxLength = -1;
//         for(int i = 1; i < n; i++){
//             LIS[i] = 1;
//             for(int j = 0; j < i; j++){
//                 if(nums[j] < nums[i]){
//                     int currLIS = LIS[j]+1;
//                     if(currLIS > LIS[i])
//                     {
//                         LIS[i] = currLIS;
                        
//                     }
//                     //LIS[i] = Math.max(LIS[i], LIS[j]+1);
//                 }
//             }
//             maxLength = Math.max(maxLength,LIS[i]);
//         }

//         int ans = 0;

//         for(int i = 0; i < n; i++){
//             if(maxLength == LIS[i])
//                 ans++;
//         }

//         return ans;
//     }
// }


class Solution {
    public int findNumberOfLIS(int[] nums) {
        int n = nums.length;
        if (n == 0) return 0;
        
        int[] LIS = new int[n]; // Length of LIS ending at each index
        int[] count = new int[n]; // Number of LIS ending at each index
        Arrays.fill(LIS, 1);
        Arrays.fill(count, 1);

        int maxLength = 1;
        for (int i = 1; i < n; i++) {
            for (int j = 0; j < i; j++) {
                if (nums[j] < nums[i]) {
                    if (LIS[j] + 1 > LIS[i]) {
                        LIS[i] = LIS[j] + 1;
                        count[i] = count[j]; // reset the count
                    } else if (LIS[j] + 1 == LIS[i]) {
                        count[i] += count[j]; // add the count
                    }
                }
            }
            maxLength = Math.max(maxLength, LIS[i]);
        }

        int ans = 0;
        for (int i = 0; i < n; i++) {
            if (LIS[i] == maxLength) {
                ans += count[i];
            }
        }

        return ans;
    }
}

File: 0673-number-of-longest-increasing-subsequence/0673-number-of-longest-increasing-subsequence.cpp
class Solution {
public:
    int findNumberOfLIS(vector<int>& nums) {
        int n = nums.size();
        if (n == 0) return 0; // Edge case: if nums is empty, return 0.

        vector<int> LIS(n, 1);   // LIS[i] stores the length of the LIS ending at index i.
        vector<int> count(n, 1); // count[i] stores the number of LIS ending at index i.

        int maxLength = 1; // Stores the maximum length of LIS found.

        // Iterate over each element in nums.
        for (int i = 1; i < n; i++) {
            for (int j = 0; j < i; j++) {
                if (nums[j] < nums[i]) { // Check if we can extend the LIS
                    if (LIS[j] + 1 > LIS[i]) { // Found a longer LIS ending at i
                        LIS[i] = LIS[j] + 1;
                        count[i] = count[j]; // Reset count[i] to count[j]
                    } else if (LIS[j] + 1 == LIS[i]) { // Another LIS of the same length
                        count[i] += count[j]; // Add count[j] to count[i]
                    }
                }
            }
            maxLength = max(maxLength, LIS[i]); // Update the max LIS length found
        }

        int ans = 0; // Stores the total number of LIS
        for (int i = 0; i < n; i++) {
            if (LIS[i] == maxLength) { // Sum counts of all LIS with maxLength
                ans += count[i];
            }
        }

        return ans;
    }
};

